src/util/mpscfifo.h

```diff
@@ -1,66 +1,88 @@
 #pragma once
 
 #include <QAtomicInt>
+#include <QMutexLocker>
 
 #include "util/assert.h"
+#include "util/memory.h"
 
-// Lock-free FIFO for multiple producers/writers and a single(!) consumer/reader.
+enum class MpscFifoConcurrency {
+    SingleProducer,
+    MultipleProducers,
+};
+
+// FIFO for multiple producers/writers and a single consumer/reader. Reading
+// is lock-free while concurrent writers are synchronized by a mutex. The
+// mutex can be disabled by explicitly creating a single producer instance.
```

Comment:
I think we need here some words how to adjust the capavity, and what does it mran. The +1 can be supprising I guess. Maybe we can also go for -1 to set number of tokens.

Reply:
I agree, the naming of both members is an open issue. Any ideas?

I wanted to achieve that the value of each atomic value never gets below `0`. By starting the reader tokens at `capacity` instead of `0` you also get a nice symmetry between acquire (+1) and release (-1) memory ordering semantics.

Reply:
I just mean that we need a comment how to set the capacity.

Looking at the member variable names, you may be right that there can be approved, because they are not original tokens like cookies in a tin. 

I will have a look.

Reply:
The capacity is fixed at compile time and cannot be changed at runtime.

Reply:
Idea:
Invert the logic:
m_writeTokens(0)  -> m_writePlaces(capacity)
m_readTokens(capacity) ->m_readPlaces(0)

Without a use case it is hard to check the requirements for a multi writer solution.

Why we need such a big control structure, only path a single pointer between two threads? 
Do you plan to use it for multi producers? Why we need an extra place. 

If the assumption fails, we have a CPU heating loop waiting for a free place. 
Does it fail if the reader is stalled for some reasons?

Reply:
*m_writeTokens/m_readTokens*
I've renamed "Tokens" to "Count" and added a comment about the initial values. Inverting the bookkeeping does not help.

*CPU heating loop*
`TrackAnalysisScheduler` will never enter a busy waiting loop in this use case, because the targeted `AnalyzerThread` became _idle_ before this code is executed. I raised the log level to _critical_ if this should ever happen unexpectedly. The code is also guarded by VERIFY_AND_DEBUG_ASSERT.

*big control structure*
`MpscFifo` is just a workaround for `std::atomic<TrackPointer>` that is simply not available. I added a note to the corresponding member variable in `AnlyzerThread`. Using the plain `TrackPointer` is almost safe in this very special use case, but it cannot be guaranteed. While idling the worker thread could theoretically be woken up at any time by anyone, causing a silent race condition for this member variable that would go unnoticed.

Extending the FIFO from a single to multiple producers comes almost for free with negligible runtime overhead, so why don't account for this? It also might not be obvious if we ever revisit this class in the future. Now it's just there for documentation.